Let users set QC status for multiple samples from /groups

alkc opened the pull request:
This update adds a dropdown menu to /groups allowing users to toggle QC status for multiple samples. 

![bonsai-qc-pr](https://github.com/Clinical-Genomics-Lund/bonsai/assets/13668555/6dcbe17a-efb1-4e12-ae9b-4e1abcb41ffe)

The QC form is submitted to new route `groups.qc_status` which updates the QC status for the selected samples and redirects back to `/groups`. 

The sample selection in the sample JS table is retained on page reload through `GET` args. This is implemented with the assumption that the end-user won't try to set the QC status for so many samples at the same time that the `GET` request grows too long.

Other changes:
* Added route `groups.qc_status`
* Added Enum for QC statuses
* Added option to /groups view to render the page with pre-selected samples in the JS table using `GET` args.
* Changed whitespace control to jinja to trim whitespace left behind by in-line template code 

Closes #97 

### Suggested test sequence:

1. Setup new bonsai instance from this branch and throw a few samples at it
2. Go to /groups and verify that the QC button is disabled
3. Select a few samples and verify that the QC controls can be accessed
4. Setting QC to passed results in a page reload and a change in the QC status of the samples
5. Setting QC to failed results in a page reload and a change in the QC status of the samples. The correct follow-up action and comment is added to the affected samples. 

### TODO:

- [ ] Review current changes
- [ ] Update CHANGELOG
- [ ] Make QC dropdown menu wider
- [ ] Organize new JS code better in new templates?
- [x] Fix merge conflicts

alkc commented on frontend/app/blueprints/groups/templates/shared.html:
```diff
@@ -113,4 +113,169 @@
+            - bad_qc_actions : List[str]    List of selectable follow-up actions for failed samples
+
+    #}
+    <div id="find-similar-dropdown" class="dropdown">
```
```suggestion
    <div id="qc-menu-dropdown" class="dropdown">
```

mhkc commented on frontend/app/app.py:
```diff
@@ -24,6 +24,10 @@ def create_app():
     app.jinja_env.globals.update(zip_longest=zip_longest)
 
+    # whitespace control:
+    app.jinja_env.trim_blocks = True
```
👍

mhkc commented on frontend/app/blueprints/groups/templates/shared.html:
```diff
@@ -113,4 +113,169 @@
+        Inputs:
+            - bad_qc_actions : List[str]    List of selectable follow-up actions for failed samples
+
+    #}
```
Nice docstring

mhkc commented on frontend/app/blueprints/groups/templates/shared.html:
```diff
@@ -113,4 +113,169 @@
+      */
+     function addSelectedSamplesToQcForm(form) {
+         // Retrieve selected samples from sessionStorage
+         var selectedSamples = window.getSelectedRows()
```
```suggestion
         let selectedSamples = window.getSelectedRows()
```

Avoid using `var` when its not necessary because they can be accidentally re declared. Its often better to use `let` or `const`

mhkc commented on frontend/app/blueprints/groups/templates/shared.html:
```diff
@@ -113,4 +113,169 @@
+    </div>
+    <script>
+     // Add data from JS table to qc bulk edit form before submit
+     let qcForm = document.getElementById("qc-form-control");
```
```suggestion
     const qcForm = document.getElementById("qc-form-control");
```

mhkc commented on frontend/app/blueprints/groups/templates/shared.html:
```diff
@@ -113,4 +113,169 @@
+
+         // Remove existing hidden inputs with the name 'qc-selected-samples'
+         var existingInputs = form.querySelectorAll('input[name="qc-selected-samples"]');
+         existingInputs.forEach(function(existingInput) {
```
```suggestion
         existingInputs.forEach(existingInput => {
```

mhkc commented on frontend/app/blueprints/groups/templates/shared.html:
```diff
@@ -113,4 +113,169 @@
+         });
+
+         // Iterate over the selectedSamples array and add as hidden inputs to the form
+         selectedSamples.forEach(function(sampleId) {
```
```suggestion
         selectedSamples.forEach(sampleId => {
```

mhkc commented on frontend/app/blueprints/groups/templates/shared.html:
```diff
@@ -113,4 +113,169 @@
+        >
+            <div id="qc-dropdown-contents" class="dropdown-menu dropdown-menu-end p-2">
+                <div class ="row pb-2">
+                    <div class="btn-group btn-group-sm col-auto" role="group" aria-label="Sample qc rejection group">
```
Can the code for the passed/ failed QC, rejection reasons and comment be shared by the sample sidebar?

alkc replied:
Maybe. The sidebar macro requires sample-specific data, but maybe that can be worked around somehow.

I'll give it a try.

alkc replied:
Done! See new comments in main PR thread.

mhkc reviewed (changes requested):
Looks good. Only a few comments. The PR needs to be rebased prior to merge.

alkc commented:
I've deduplicated all the QC form code back into `sample/sidebar.html`. Could be moved to some non-blueprint specific template dir, but I'm saving that for some future update. 

the /samples sidebar and /groups QC form now use the same macros for both the HTML and JS enhancements.

### Demo of QC in /groups:

![bonsai-qc-pr-update1](https://github.com/Clinical-Genomics-Lund/bonsai/assets/13668555/e26e63e0-3895-4481-a65e-dee61a91187c)

### Demo of QC in /samples

![bonsai-qc-pr-update2](https://github.com/Clinical-Genomics-Lund/bonsai/assets/13668555/1692b634-4714-43d7-9b0b-f84b77b237ed)

alkc commented:
I've rebased the branch onto master (hopefully w/o breaking anything), incorporated the suggested fixes and deduplicated the QC form templates/macros.

See above for updated demos.
